otherlibs/systhreads/st_stubs.c

```diff
@@ -545,6 +550,12 @@ CAMLprim value caml_thread_initialize(value unit)
     caml_failwith("caml_thread_initialize: cannot initialize Thread "
                   "while several domains are running.");
 
+  thread_table = caml_stat_calloc_noexc(caml_params->max_domains,
+                                        sizeof(struct caml_thread_table));
+  if (thread_table == NULL)
+    caml_fatal_error("caml_thread_initialize: failed to allocate thread"
```

Comment:
At first I was concerned that the previous failure point in this function uses `caml_failwith`, but it can only occur in a specific cinematic where it makes sense to raise an exception, while these added allocation failures should be considered fatal (hence `caml_fatal_error`).

Eventually (in a later PR) this routine should be allowed to return a `caml_result`, but that's more work (and orthogonal to the problem being solved here).

Reply:
Proper error handling in this function is actually part of #12410.

(Incidentally I noticed that this PR interferes with many others: #12410, #12964, #13026.)

Reply:
Actually, I think these fatal errors should be `Out_of_memory` exceptions instead, since the `caml_failwith` above shows that it is admissible to fail with uninitialized thread machinery.

Now this raises the question of whether the `caml_failwith` above is sound. It is sound if we are sure that an uncaught exception during module initialization stops the program (otherwise we need hardening by checking the variable `threads_initialized` at the entry points of the Thread module). Unfortunately I don't know enough about module initialization to answer this question.

Reply:
Is it a good idea to continue program execution when `caml_stat_calloc_noexc` fails to allocate memory for `thread_table`? Given that the underlying `calloc` has failed, the program is very likely doomed to fail soon in a non-recoverable way. Better to hard fail (fatal error) as soon as we anticipate this.

Like @gadmm, I also have the question of whether `caml_failwith` is sound. In particular, the exception that may be raised by this function is not handled at the call site,

https://github.com/ocaml/ocaml/blob/6ad156a735839b367fd2841bb3f6205316d9be4b/otherlibs/systhreads/thread.ml#L76-L79

and will lead to module initialisation failing with an uncaught exception?

Reply:
My reasoning here is that the Out_of_memory will halt the program quickly anyway. Indeed, otherwise there is also an issue right above with `caml_failwith`. This can provide better debugging. Since this is a large allocation (e.g. can be above glibc MMAP_THRESHOLD), this does not mean that the remaining mallocs are going to fail.

But I agree that the main concern is the Failure above, which happens if another module spawns domains during initialization. Do we know if, when module initialization fails:
- no functions from this module can be called?
- the program stops?

Reply:
The semantics I *expect* if the evaluation of a module raises an exception is that the whole program fails with that exception, without executing the code of later modules in link order. (The mental model of `ocamlc a.ml b.ml c.ml` is that the produced program performs the sequential evaluation of modules A, B, C.)

However, taking a step back: I think this part of the runtime is an unfinished design, not something set in stone that we should think deeply about. We should admit that the runtime model exposed to our users is a small number of domains with arbitrarily many non-parallel threads per domain, and move whatever parts of the thread machinery into the `runtime/` logic that let us do this in a simpler, more robust way. It's "just" that no one around is actively working on this.